### PR TITLE
:sparkles: Add MachineDrainRule "WaitCompleted"

### PR DESCRIPTION
**Pull request by vincepri**
**What this PR does / why we need it**:

This PR adds the ability for drain to wait for the completion of specific pods. This is useful in scenario where drain is either handled outside the context of `kubectl drain` after a Node is cordoned, or for long running batch Jobs that should be allowed to terminate on their own.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #

### CODE REVIEW

File: api/v1beta1/machinedrainrules_types.go [hunk 1]
```diff
@@ -22,13 +22,13 @@ import (
...
-	// The only valid value is "skip".
+	// The only valid value is "skip" or "waitcompleted".
 	// This label takes precedence over MachineDrainRules defined in the management cluster.
 	PodDrainLabel = "cluster.x-k8s.io/drain"
 )
 
-// MachineDrainRuleDrainBehavior defines the drain behavior. Can be either "Drain" or "Skip".
-// +kubebuilder:validation:Enum=Drain;Skip
+// MachineDrainRuleDrainBehavior defines the drain behavior. Can be either "Drain", "Skip", or "WaitCompleted".
+// +kubebuilder:validation:Enum=Drain;Skip;WaitCompleted
```
Comment (sbueringer) [inline api/v1beta1/machinedrainrules_types.go:31]: Please check if the proposal should be amended accordingly

File: api/v1beta1/machinedrainrules_types.go [hunk 2]
```diff
@@ -22,13 +22,13 @@ import (
 
 const (
 	// PodDrainLabel is the label that can be set on Pods in workload clusters to ensure a Pod is not drained.
-	// The only valid value is "skip".
+	// The only valid value is "skip" or "waitcompleted".
```
Comment (sbueringer) [inline api/v1beta1/machinedrainrules_types.go:25]:
```suggestion
	// The only valid values are "skip" and "waitcompleted".
```

File: internal/controllers/machine/drain/drain.go [hunk 3]
```diff
@@ -498,6 +507,10 @@ func (r EvictionResult) ConditionMessage(nodeDrainStartTime *metav1.Time) string
 		conditionMessage = fmt.Sprintf("%s\nAfter above Pods have been removed from the Node, the following Pods will be evicted: %s",
 			conditionMessage, PodListToString(r.PodsToTriggerEvictionLater, 3))
 	}
+	if len(r.PodsToWaitCompleted) > 0 {
+		conditionMessage = fmt.Sprintf("%s\nWaiting for the following Pods to complete without drain: %s",
```
Comment (sbueringer) [inline internal/controllers/machine/drain/drain.go:511]:
```suggestion
		conditionMessage = fmt.Sprintf("%s\nWaiting for the following Pods to complete without eviction: %s",
```
Maybe more precise

File: internal/controllers/machine/drain/filters.go [hunk 4]
```diff
@@ -275,6 +286,11 @@ func (d *Helper) drainLabelFilter(ctx context.Context, pod *corev1.Pod) PodDelet
 		log.V(4).Info(fmt.Sprintf("Skip evicting Pod, because Pod has %s label", clusterv1.PodDrainLabel))
 		return MakePodDeleteStatusSkip()
 	}
+	if labelValue, found := pod.ObjectMeta.Labels[clusterv1.PodDrainLabel]; found && strings.EqualFold(labelValue, string(clusterv1.MachineDrainRuleDrainBehaviorWaitCompleted)) {
+		log := ctrl.LoggerFrom(ctx, "Pod", klog.KObj(pod))
+		log.V(4).Info(fmt.Sprintf("Skip evicting Pod, because Pod has %s label", clusterv1.PodDrainLabel))
```
Comment (sbueringer) [inline internal/controllers/machine/drain/filters.go:291]: Should we maybe start mentioning the label value (here and in l.286)

File: internal/controllers/machine/drain/filters.go [hunk 5]
```diff
@@ -300,6 +316,8 @@ func (d *Helper) machineDrainRulesFilter(machineDrainRules []*clusterv1.MachineD
 				log := ctrl.LoggerFrom(ctx, "Pod", klog.KObj(pod))
 				log.V(4).Info(fmt.Sprintf("Skip evicting Pod, because MachineDrainRule %s with behavior %s applies to the Pod", mdr.Name, clusterv1.MachineDrainRuleDrainBehaviorSkip))
 				return MakePodDeleteStatusSkip()
+			case clusterv1.MachineDrainRuleDrainBehaviorWaitCompleted:
```
Comment (sbueringer) [inline internal/controllers/machine/drain/filters.go:319]: Let's add the same log as in l.316+317

File: internal/controllers/machine/drain/drain.go [hunk 6]
```diff
@@ -281,6 +281,7 @@ func (d *Helper) EvictPods(ctx context.Context, podDeleteList *PodDeleteList) Ev
 	var podsToTriggerEvictionLater []PodDelete
 	var podsWithDeletionTimestamp []PodDelete
```
Comment (sbueringer) [inline internal/controllers/machine/drain/drain.go:282]: Please check the existing test coverage for the modified funcs and extend accordingly for the new case (we should have test coverage for everything, only new cases should be needed)

File: internal/controllers/machine/drain/drain.go [hunk 7]
```diff
@@ -498,6 +507,10 @@ func (r EvictionResult) ConditionMessage(nodeDrainStartTime *metav1.Time) string
 		conditionMessage = fmt.Sprintf("%s\nAfter above Pods have been removed from the Node, the following Pods will be evicted: %s",
 			conditionMessage, PodListToString(r.PodsToTriggerEvictionLater, 3))
 	}
+	if len(r.PodsToWaitCompleted) > 0 {
```
Comment (sbueringer) [inline internal/controllers/machine/drain/drain.go:510]: @fabriziopandini Is this something that should be specifically handled on higher levels when the condition bubbles up? (like we handled PDB's etc.)

(could also be a follow-up PR)

File: internal/controllers/machine/drain/filters.go [hunk 8]
```diff
@@ -275,6 +286,11 @@ func (d *Helper) drainLabelFilter(ctx context.Context, pod *corev1.Pod) PodDelet
 		log.V(4).Info(fmt.Sprintf("Skip evicting Pod, because Pod has %s label", clusterv1.PodDrainLabel))
```
Comment (sbueringer) [inline internal/controllers/machine/drain/filters.go:286]: let's update the godocs in drain.go l.131 + l.153

File: internal/controllers/machine/drain/filters.go [hunk 9]
```diff
@@ -275,6 +286,11 @@ func (d *Helper) drainLabelFilter(ctx context.Context, pod *corev1.Pod) PodDelet
...
 		return MakePodDeleteStatusSkip()
 	}
+	if labelValue, found := pod.ObjectMeta.Labels[clusterv1.PodDrainLabel]; found && strings.EqualFold(labelValue, string(clusterv1.MachineDrainRuleDrainBehaviorWaitCompleted)) {
```
Comment (sbueringer) [inline internal/controllers/machine/drain/filters.go:289]: I think we also have to update drain.go l.228 to break for MachineDrainRuleDrainBehaviorWaitCompleted. Otherwise the filter result will be overwritten by subsequent filters (in this case the machineDrainRulesFilter)

File: internal/controllers/machine/drain/drain.go (hunk 6 above)
Comment (sbueringer) [reply internal/controllers/machine/drain/drain.go:282]: Let's also extend the NodeDrainTimeout e2e test to cover waitcompleted (node_drain.go)

(Unfortunately the whole drain implementation is a bit brittle (already before this PR) so e2e coverage would be really good. For example with the current PR even Pods with the label set to waitcompleted would have been drained because the machineDrainRulesFilter would have overwritten the result of the drainLabelFilter with behavior drain)

Comment (vincepri) [reply internal/controllers/machine/drain/drain.go:283]: done both

File: internal/controllers/machine/drain/drain.go (hunk 7 above)
Comment (fabriziopandini) [reply internal/controllers/machine/drain/drain.go:515]: Yes, you are correct, we have to handle the new use case in:
-https://github.com/kubernetes-sigs/cluster-api/blob/d8fd3b57fae20b92c190e43798ac35ec4b2a7dd5/internal/controllers/machine/machine_controller_status.go#L702-L716
- https://github.com/kubernetes-sigs/cluster-api/blob/c333db6c2303aa6cc30e1aefb8b31d49ef20acd3/internal/controllers/machineset/machineset_controller_status.go#L442-L462 and similar func for MD, KCP

Comment (sbueringer) [reply internal/controllers/machine/drain/drain.go:515]: @fabriziopandini This PR or follow-up? I lean towards follow-up, mostly just to not expand the scope of this PR further

File: internal/controllers/machine/drain/drain.go [hunk 10]
```diff
@@ -498,6 +512,10 @@ func (r EvictionResult) ConditionMessage(nodeDrainStartTime *metav1.Time) string
 		conditionMessage = fmt.Sprintf("%s\nAfter above Pods have been removed from the Node, the following Pods will be evicted: %s",
 			conditionMessage, PodListToString(r.PodsToTriggerEvictionLater, 3))
 	}
```
Comment (sbueringer) [inline internal/controllers/machine/drain/drain.go:514]: I think we should move the if for PodsToWaitCompleted above the one for PodsToTriggerEvictionLater

The message for PodsToTriggerEvictionLater contains "After above Pods ..." and this only makes sense if a potential "Waiting for the following Pods..." message is actually above

File: internal/controllers/machine/drain/drain_test.go [hunk 11]
```diff
@@ -1356,6 +1441,202 @@ func TestEvictPods(t *testing.T) {
...
+				},
+				{
+					Pod: &corev1.Pod{
+						ObjectMeta: metav1.ObjectMeta{
+							Name: "pod-8-to-trigger-eviction-later",
+						},
+					},
+					Status: PodDeleteStatus{
+						DrainBehavior: clusterv1.MachineDrainRuleDrainBehaviorDrain,
+						DrainOrder:    ptr.To[int32](6), // min DrainOrder is 6 => will be evicted later
```
Comment (sbueringer) [inline internal/controllers/machine/drain/drain_test.go:1552]:
```suggestion
						DrainOrder:    ptr.To[int32](6), // min DrainOrder is -5 => will be evicted later
```
nit: this is referring to the calculated min order of all Pods

File: internal/controllers/machine/drain/drain.go (hunk 10 above)
Comment (sbueringer) [reply internal/controllers/machine/drain/drain.go:514]: Hm okay this is a bit more complicated.

The idea is basicallly that this condition message surfaces where the drain is currently stuck (and which Pods we are going to evict next).

The problem with the current behavior is that "WaitCompleted"-Pods are sometimes blocking further progress and sometimes they don't (it depends on the order of the Pods that we have to evict).

I would suggest to treat WaitCompleted Pods similar to PodsToTriggerEviction.

This means splitting them up into `podsToWaitCompletedNow` and `podsToWaitCompletedLater` in `EvictPods` (based on if `ptr.Deref(pod.Status.DrainOrder, 0) == minDrainOrder`)

I would then also update the log statement in l.303 accordingly.

Once that is done we can also write the condition message here accordingly. I would suggest:
```go
	if len(r.PodsToWaitCompletedNow) > 0 {
		kind := "Pod"
		if len(r.PodsToWaitCompletedNow) > 1 {
			kind = "Pods"
		}
		// Note: the code computing stale warning for the machine deleting condition is making assumptions on the format/content of this message.
		// Same applies for other conditions where deleting is involved, e.g. MachineSet's Deleting and ScalingDown condition.
		conditionMessage = fmt.Sprintf("%s\n* %s %s: waiting for completion",
			conditionMessage, kind, PodListToString(r.PodsDeletionTimestampSet, 3))
	}	
	
	if len(r.PodsToTriggerEvictionLater) > 0 {
		conditionMessage = fmt.Sprintf("%s\nAfter above Pods have been removed from the Node, the following Pods will be evicted: %s",
			conditionMessage, PodListToString(r.PodsToTriggerEvictionLater, 3))
	}

	if len(r.PodsToWaitCompletedLater) > 0 {
		conditionMessage = fmt.Sprintf("%s\nAfter above Pods have been removed from the Node, waiting for the following Pods to complete without eviction: %s",
			conditionMessage, PodListToString(r.PodsToWaitCompletedLater, 3))
	}
```

File: test/e2e/node_drain.go [hunk 12]
```diff
@@ -233,6 +233,31 @@ func NodeDrainTimeoutSpec(ctx context.Context, inputGetter func() NodeDrainTimeo
...
 		}
+		By("Deploy an unevictable Deployment with `wait-completed` label.")
+		framework.DeployPodAndWait(ctx, framework.DeployPodAndWaitInput{
+			WorkloadClusterProxy: workloadClusterProxy,
+			ControlPlane:         controlplane,
+			DeploymentName:       cpDeploymentName("wait-completed"),
+			Namespace:            "unevictable-workload",
+			NodeSelector:         map[string]string{nodeOwnerLabelKey: "KubeadmControlPlane-" + controlplane.Name},
+			ModifyDeployment: func(deployment *appsv1.Deployment) {
+				deployment.Spec.Template.Labels["cluster.x-k8s.io/drain"] = string(clusterv1.MachineDrainRuleDrainBehaviorWaitCompleted)
```
Comment (sbueringer) [inline test/e2e/node_drain.go:244]:
```suggestion
				deployment.Spec.Template.Labels["cluster.x-k8s.io/drain"] = "wait-completed"
```
nit: I know both works, but maybe let's test the one that we document in the API (same in l.256)

File: test/e2e/node_drain.go [hunk 13]
```diff
@@ -207,7 +207,7 @@ func NodeDrainTimeoutSpec(ctx context.Context, inputGetter func() NodeDrainTimeo
 		}
 
 		By("Deploy Deployment with unevictable Pods on control plane and MachineDeployment Nodes.")
```
Comment (sbueringer) [inline test/e2e/node_drain.go:209]: Please update the test docs in l.79++

This should be

```go
// NodeDrainTimeoutSpec goes through the following steps:
// * Create cluster with 3 CP & 1 worker Machine
// * Ensure Node label is set & NodeDrainTimeout is set to 0 (wait forever)
// * Deploy MachineDrainRules
// * Deploy Deployment with unevictable Pods on CP & MD Nodes
// * Deploy Deployment with unevictable Pods with `wait-completed` label on CP & MD Nodes
// * Deploy Deployment with evictable Pods with finalizer on CP & MD Nodes
// * Deploy additional resources if defined in input
// * Trigger Node drain by scaling down the control plane to 1 and MachineDeployments to 0
// * Get draining control plane and MachineDeployment Machines
// * Verify drain of Deployments with order 1
// * Verify drain of Deployments with order 5
// * Verify skipped Pods are still there and don't have a deletionTimestamp
// * Verify wait-completed Pods are still there and don't have a deletionTimestamp
// * Verify Node drains for control plane and MachineDeployment Machines are blocked (by PDBs and WaitCompleted Pods)
// * Set NodeDrainTimeout to 1s to unblock Node drain
// * Verify machine deletion is blocked by waiting for volume detachment (only if VerifyNodeVolumeDetach is enabled)
// * Unblocks waiting for volume detachment (only if VerifyNodeVolumeDetach is enabled)
// * Verify scale down succeeded because Node drains were unblocked.
```

(we'll have to change the order values, see comment below)

File: test/e2e/node_drain.go [hunk 14]
```diff
@@ -412,6 +437,26 @@ func NodeDrainTimeoutSpec(ctx context.Context, inputGetter func() NodeDrainTimeo
...
+			Expect(workloadClusterProxy.GetClient().List(ctx, skippedMDPods,
+				client.InNamespace("unevictable-workload"),
+				client.MatchingLabels{"deployment": mdDeploymentName("wait-completed", md.Name)},
+				client.MatchingFieldsSelector{Selector: fields.OneTermEqualSelector("spec.nodeName", drainingMDNodeNames[md.Name])},
+			)).To(Succeed())
+			Expect(skippedMDPods.Items).To(HaveLen(1))
+			Expect(skippedMDPods.Items[0].DeletionTimestamp.IsZero()).To(BeTrue())
+		}
+
 		By("Verify Node drains for control plane and MachineDeployment Machines are blocked (only by PDBs)")
```
Comment (sbueringer) [inline test/e2e/node_drain.go:460]:
```suggestion
		By("Verify Node drains for control plane and MachineDeployment Machines are blocked (by PDBs and WaitCompleted Pods)")
```

File: test/e2e/node_drain.go [hunk 15]
```diff
@@ -233,6 +233,31 @@ func NodeDrainTimeoutSpec(ctx context.Context, inputGetter func() NodeDrainTimeo
 				WaitForDeploymentAvailableInterval: input.E2EConfig.GetIntervals(specName, "wait-deployment-available"),
 			})
 		}
+		By("Deploy an unevictable Deployment with `wait-completed` label.")
```
Comment (sbueringer) [inline test/e2e/node_drain.go:236]: I think in general the test can work roughly as is.

But as we are now assuming order 0 for WaitCompleted Pods we have to change the order of some other Pods to make sure they can be drained (currently the drain of these Pods is blocked by the waitcompleted Pods).

tl;dr I think the Pods with drain order 1 should get order -5, the Pods with drain order 5 should get order -1

Then all of these Pods can be evicted / drained

File: test/framework/deployment_helpers.go [hunk 16]
```diff
@@ -552,9 +528,36 @@ func DeployUnevictablePod(ctx context.Context, input DeployUnevictablePodInput)
...
 		Namespace: input.Namespace,
-		ClientSet: workloadClient,
+		ClientSet: input.WorkloadClusterProxy.GetClientSet(),
 		Budget:    budget,
 	})
+}
+
+// DeployPodAndWait will deploy a Deployment on a ControlPlane or MachineDeployment.
+func DeployPodAndWait(ctx context.Context, input DeployPodAndWaitInput) {
+	Expect(input.DeploymentName).ToNot(BeNil(), "Need a deployment name in DeployUnevictablePod")
```
Comment (sbueringer) [inline test/framework/deployment_helpers.go:538]:
```suggestion
	Expect(input.DeploymentName).ToNot(BeNil(), "Need a deployment name in DeployPodAndWait")
```
nit, same in the few lines below

File: test/e2e/node_drain.go (hunk 15 above)
Comment (vincepri) [reply test/e2e/node_drain.go:236]: done, ptal

File: internal/controllers/machine/drain/drain.go (hunk 10 above)
Comment (vincepri) [reply internal/controllers/machine/drain/drain.go:533]: done, ptal

File: internal/controllers/machine/drain/drain.go [hunk 17]
```diff
@@ -431,13 +451,16 @@ type EvictionResult struct {
...
+	PodsToWaitCompletedNow     []*corev1.Pod
+	PodsToWaitCompletedLater   []*corev1.Pod
 	PodsNotFound               []*corev1.Pod
 	PodsIgnored                []*corev1.Pod
 }
 
 // DrainCompleted returns if a Node is entirely drained, i.e. if all relevant Pods have gone away.
 func (r EvictionResult) DrainCompleted() bool {
-	return len(r.PodsDeletionTimestampSet) == 0 && len(r.PodsFailedEviction) == 0 && len(r.PodsToTriggerEvictionLater) == 0
+	return len(r.PodsDeletionTimestampSet) == 0 && len(r.PodsFailedEviction) == 0 &&
```
Comment (sbueringer) [inline internal/controllers/machine/drain/drain.go:462]: I think this should also include PodsToWaitCompletedNow
(might explain the e2e test failure)

**Comment by sbueringer** [comment]
@vincepri One finding + `make generate-go-openapi`

Unit test can be ignored, has nothing to do with your PR and I already opened an issue for that one

**Comment by enxebre** [comment]
one question https://github.com/kubernetes-sigs/cluster-api/pull/11545/files#r1924079699
lgtm otherwise

File: internal/controllers/machine/drain/filters.go [hunk 18]
```diff
@@ -272,9 +284,15 @@ func (d *Helper) skipDeletedFilter(ctx context.Context, pod *corev1.Pod) PodDele
...
 	if labelValue, found := pod.ObjectMeta.Labels[clusterv1.PodDrainLabel]; found && strings.EqualFold(labelValue, string(clusterv1.MachineDrainRuleDrainBehaviorSkip)) {
 		log := ctrl.LoggerFrom(ctx, "Pod", klog.KObj(pod))
-		log.V(4).Info(fmt.Sprintf("Skip evicting Pod, because Pod has %s label", clusterv1.PodDrainLabel))
+		log.V(4).Info(fmt.Sprintf("Skip evicting Pod, because Pod has %s label with %s value", clusterv1.PodDrainLabel, labelValue))
 		return MakePodDeleteStatusSkip()
 	}
+	if labelValue, found := pod.ObjectMeta.Labels[clusterv1.PodDrainLabel]; found &&
+		strings.EqualFold(strings.Replace(labelValue, "-", "", 1), string(clusterv1.MachineDrainRuleDrainBehaviorWaitCompleted)) {
+		log := ctrl.LoggerFrom(ctx, "Pod", klog.KObj(pod))
+		log.V(4).Info(fmt.Sprintf("Skip evicting Pod, because Pod has %s label with %s value", clusterv1.PodDrainLabel, labelValue))
```
[the remark this answers is not in the document]
Comment (vincepri) [reply internal/controllers/machine/drain/filters.go:293]: Will add a log line, not sure about surfacing it up in the condition at this point, could be a follow-up

**Comment by sbueringer** [comment]
Thank you!

/lgtm

/assign @fabriziopandini @chrischdi 

/test pull-cluster-api-e2e-main

File: internal/controllers/machine/drain/drain.go (hunk 7 above)
Comment (fabriziopandini) [reply internal/controllers/machine/drain/drain.go]: sgtm

**Comment by fabriziopandini** [comment]
/lgtm
If we plan to backport, let's make sure we backport when also https://github.com/kubernetes-sigs/cluster-api/pull/11545#discussion_r1899697663 is addressed in a follow up PR

**Comment by sbueringer** [comment]
@vincepri Do you have time to open a PR for: https://github.com/kubernetes-sigs/cluster-api/pull/11545#discussion_r1899697663?
